diff --git a/download.c b/download.c
index 0417f4c..6e2a34f 100644
--- a/download.c
+++ b/download.c
@@ -263,11 +263,6 @@ int apply_downloaded_file(struct cwmp *cwmp, struct download *pdownload, struct
 		if (cwmp_apply_firmware() != 0)
 			error = FAULT_CPE_DOWNLOAD_FAIL_FILE_CORRUPTED;

-		if (error == FAULT_CPE_NO_FAULT) {
-			sleep(70);
-			error = FAULT_CPE_DOWNLOAD_FAIL_FILE_CORRUPTED;
-		}
-
 	} else if (strcmp(pdownload->file_type, WEB_CONTENT_FILE_TYPE) == 0) {
 		//TODO Not Supported
 		error = FAULT_CPE_NO_FAULT;
