diff --git a/notifications.c b/notifications.c
index 0c95a6e..29a859f 100644
--- a/notifications.c
+++ b/notifications.c
@@ -34,7 +34,6 @@ char *notifications[7] = {"disabled" , "passive", "active", "passive_lw", "passi
 
 char *default_active_notifications_parameters[] = {
 	"Device.ManagementServer.ConnectionRequestURL",
-	"Device.ManagementServer.ConnReqJabberID",
 	"Device.GatewayInfo.ManufacturerOUI",
 	"Device.GatewayInfo.ProductClass",
 	"Device.GatewayInfo.SerialNumber",
