diff --git a/download.c b/download.c
index a6cf59f..31b187c 100644
--- a/download.c
+++ b/download.c
@@ -328,7 +328,7 @@ int apply_downloaded_file(struct cwmp *cwmp, struct download *pdownload, char *d
 		if (pdownload->file_type[0] == '3') {
 			CWMP_LOG(INFO, "Download and apply new vendor config file is done successfully");
 		}
-		return FAULT_CPE_NO_FAULT;
+		CWMP_LOG(INFO, "Applied downloaded image successfully ....");
 	}
 	if (error != FAULT_CPE_NO_FAULT) {
 		bkp_session_delete_transfer_complete(ptransfer_complete);
