diff --git a/ubus_utils.c b/ubus_utils.c
index c21576c..9244207 100644
--- a/ubus_utils.c
+++ b/ubus_utils.c
@@ -187,6 +187,13 @@ static void bb_add_icwmp_status(struct blob_buf *bb)
 	global_string_param_read(&cwmp_main.conf.acsurl, &acs_url);
 	bb_add_string(bb, "acs_url", acs_url);
 	FREE(acs_url);
+
+	char connection_request_url[256] = { 0 };
+	char *connection_request_ip = NULL;
+	global_string_param_read(&cwmp_main.conf.connection_request_path, &connection_request_ip);
+	sprintf(connection_request_url, "http://%s:%d/", connection_request_ip, cwmp_main.conf.connection_request_port);
+	bb_add_string(bb, "connection_request_url", connection_request_url);
+	FREE(connection_request_ip);
 	blobmsg_close_table(bb, tbl);
 }
 
