Index: icwmp-8.4.12/rpc_soap.c
===================================================================
--- icwmp-8.4.12.orig/rpc_soap.c
+++ icwmp-8.4.12/rpc_soap.c
@@ -65,10 +65,15 @@ char *forced_inform_parameters[] = {
 	"Device.RootDataModelVersion",
 	"Device.DeviceInfo.HardwareVersion",
 	"Device.DeviceInfo.SoftwareVersion",
+	"Device.DeviceInfo.ManufacturerOUI",
+	"Device.DeviceInfo.ModelName",
+	"Device.DeviceInfo.ProductClass",
+	"Device.DeviceInfo.SerialNumber",
 	"Device.DeviceInfo.ProvisioningCode",
 	"Device.ManagementServer.ParameterKey",
-	DM_CONN_REQ_URL,
-	"Device.ManagementServer.AliasBasedAddressing"
+	"Device.ManagementServer.ConnectionRequestURL",
+	"Device.ManagementServer.AliasBasedAddressing",
+	DM_CONN_REQ_URL
 };
 
 int xml_handle_message(struct session *session)
