Index: asterisk-20.3.0/res/Makefile
===================================================================
--- asterisk-20.3.0.orig/res/Makefile
+++ asterisk-20.3.0/res/Makefile
@@ -27,7 +27,7 @@ ifneq ($(findstring $(OSARCH), mingw32 c
 endif
 
 res_pjsip_pubsub.so: LIBS+=-lubus -lubox -lpicoevent
-res_pjsip_outbound_registration.so: LIBS+=-lubus -lubox -lpicoevent
+res_pjsip_outbound_registration.so: LIBS+=-lubus -lubox -lpicoevent -lindigo
 
 res_config_ldap.o: _ASTCFLAGS+=-DLDAP_DEPRECATED
 
Index: asterisk-20.3.0/res/res_pjsip_outbound_registration.c
===================================================================
--- asterisk-20.3.0.orig/res/res_pjsip_outbound_registration.c
+++ asterisk-20.3.0/res/res_pjsip_outbound_registration.c
@@ -29,6 +29,7 @@
 #include <pjsip.h>
 #include <pjsip_ua.h>
 #include <time.h>
+#include <indigo/report.h>
 
 #include "asterisk/res_pjsip.h"
 #include "asterisk/res_pjsip_cli.h"
@@ -481,6 +482,22 @@ static struct stasis_subscription *netwo
 /*! \brief Free any endpoint security mechanisms that have been stored */
 static void clear_endpoint_security_mechanisms(struct sip_outbound_registration_client_state *client_state);
 
+#define REGISTRATION_CATEGORY "voice.registration"
+static void report_registration_info(enum sip_outbound_registration_status status, const char *event, ...)
+{
+	va_list args;
+	char *buf;
+
+	va_start(args, event);
+	vasprintf(&buf, event, args);
+	va_end(args);
+
+	REPORT_LOW(REGISTRATION_CATEGORY, "{ \"status\":\"%s\", \"event\":\"%s\" }",
+		sip_outbound_registration_status_str(status), buf);
+
+	free(buf);
+}
+
 /*! \brief hashing function for state objects */
 static int registration_state_hash(const void *obj, const int flags)
 {
@@ -1628,6 +1645,8 @@ static int handle_registration_response(
 			ast_debug(1, "Outbound registration to '%s' with client '%s' successful\n", server_uri, client_uri);
 			update_client_state_status(response->client_state, SIP_REGISTRATION_REGISTERED);
 			sip_outbound_registration_send_ubus_event("REGISTERED",response->expiration,client_uri);
+			report_registration_info(response->client_state->status, "Successful registration");
+
 			response->client_state->retries = 0;
 			schedule_registration(response->client_state, next_registration_time(response->expiration));
 
@@ -1658,6 +1677,8 @@ static int handle_registration_response(
 			}
 			update_client_state_status(response->client_state, SIP_REGISTRATION_UNREGISTERED);
 			sip_outbound_registration_send_ubus_event("UNREGISTERED",response->expiration,client_uri);
+			report_registration_info(response->client_state->status, "Successful de-registration");
+
 			if (PJSIP_TRANSPORT_IS_RELIABLE(response->rdata->tp_info.transport)) {
 				ast_sip_transport_monitor_unregister_key(response->transport_key,
 					registration_transport_shutdown_cb, response->client_state->registration_name,
@@ -1679,9 +1700,12 @@ static int handle_registration_response(
 				server_uri, client_uri);
 			/* 494 loop detected! This is fatal! */
 			update_client_state_status(response->client_state, SIP_REGISTRATION_REJECTED_PERMANENT);
+			report_registration_info(response->client_state->status, "494 loop detected");
 			sip_outbound_registration_send_ubus_event("UNREGISTERED",response->expiration,client_uri);
 		} else {
 			/* Try (initial) registration again - but now with additional headers */
+			report_registration_info(response->client_state->status, "494 received");
+
 			response->client_state->is494 = 1;
 			ao2_ref(response->client_state, +1);
 			handle_client_registration(response->client_state);
@@ -1693,18 +1717,24 @@ static int handle_registration_response(
 	} else if (response->retry_after) {
 		/* If we have been instructed to retry after a period of time, schedule it as such */
 		schedule_retry(response, response->retry_after, server_uri, client_uri);
+
+		report_registration_info(response->client_state->status, "Requested to retry after %u seconds", response->retry_after);
+
 	} else if (response->client_state->retry_interval
 		&& sip_outbound_registration_is_temporal(response->code, response->client_state)) {
 		if (response->client_state->retries == response->client_state->max_retries) {
 			/* If we received enough temporal responses to exceed our maximum give up permanently */
 			update_client_state_status(response->client_state, SIP_REGISTRATION_REJECTED_PERMANENT);
 			sip_outbound_registration_send_ubus_event("UNREGISTERED",response->expiration,client_uri);
+			report_registration_info(response->client_state->status, "Maximum retries reached");
+
 			ast_log(LOG_WARNING, "Maximum retries reached when attempting outbound registration to '%s' with client '%s', stopping registration attempt\n",
 				server_uri, client_uri);
 		} else {
 			/* On the other hand if we can still try some more do so */
 			response->client_state->retries++;
 			schedule_retry(response, response->client_state->retry_interval, server_uri, client_uri);
+			report_registration_info(response->client_state->status, "Temporal failure, retry after %u seconds", response->client_state->retry_interval);
 		}
 	} else {
 		if (response->code == 403
@@ -1712,6 +1742,8 @@ static int handle_registration_response(
 			&& response->client_state->retries < response->client_state->max_retries) {
 			/* A forbidden response retry interval is configured and there are retries remaining */
 			update_client_state_status(response->client_state, SIP_REGISTRATION_REJECTED_TEMPORARY);
+			report_registration_info(response->client_state->status, "403 response, retry after %u seconds", response->client_state->forbidden_retry_interval);
+
 			sip_outbound_registration_send_ubus_event("UNREGISTERED",response->expiration,client_uri);
 			response->client_state->retries++;
 			schedule_registration(response->client_state, response->client_state->forbidden_retry_interval);
@@ -1721,6 +1753,8 @@ static int handle_registration_response(
 			   && response->client_state->retries < response->client_state->max_retries) {
 			/* Some kind of fatal failure response received, so retry according to configured interval */
 			update_client_state_status(response->client_state, SIP_REGISTRATION_REJECTED_TEMPORARY);
+			report_registration_info(response->client_state->status, "%d response, retry after %u seconds", response->code, response->client_state->fatal_retry_interval);
+
 			sip_outbound_registration_send_ubus_event("UNREGISTERED",response->expiration,client_uri);
 			response->client_state->retries++;
 			schedule_registration(response->client_state, response->client_state->fatal_retry_interval);
@@ -1730,6 +1764,7 @@ static int handle_registration_response(
 			/* Finally if there's no hope of registering give up */
 			update_client_state_status(response->client_state, SIP_REGISTRATION_REJECTED_PERMANENT);
 			sip_outbound_registration_send_ubus_event("UNREGISTERED",response->expiration,client_uri);
+			report_registration_info(response->client_state->status, "Registration failed");
 			if (response->rdata) {
 				ast_log(LOG_WARNING, "Fatal response '%d' received from '%s' on registration attempt to '%s', stopping outbound registration\n",
 					response->code, server_uri, client_uri);
