Index: asterisk-18.5.1/main/asterisk.c
===================================================================
--- asterisk-18.5.1.orig/main/asterisk.c
+++ asterisk-18.5.1/main/asterisk.c
@@ -3500,6 +3500,12 @@ int main(int argc, char *argv[])
 	struct rlimit l;
 	static const char *getopt_settings = "BC:cde:FfG:ghIiL:M:mnpqRrs:TtU:VvWXx:";
 
+	/* Configure RT priority on parent thread and all child threads */
+	struct sched_param param;
+	prctl(PR_SET_NAME, "asterisk-main", 0, 0, 0);
+	param.sched_priority = sched_get_priority_max(SCHED_FIFO);
+	pthread_setschedparam(pthread_self(), SCHED_FIFO, &param);
+
 	/* Remember original args for restart */
 	if (argc > ARRAY_LEN(_argv) - 1) {
 		fprintf(stderr, "Truncating argument size to %d\n", (int)ARRAY_LEN(_argv) - 1);
