--- a/main/cdr.c
+++ b/main/cdr.c
@@ -3636,13 +3636,24 @@ static void post_cdr(struct ast_cdr *cdr
 {
 	struct module_config *mod_cfg;
 	struct cdr_beitem *i;
+	struct ast_cdr *head_cdr = cdr;
+	int answered = 0;
 
 	mod_cfg = ao2_global_obj_ref(module_configs);
 	if (!mod_cfg) {
 		return;
 	}
+	for (; cdr ; cdr = cdr->next) {
+		if (cdr->disposition == AST_CDR_ANSWERED) {
+			ast_debug(1, "Mark the CDR as answered %s n", cdr->channel);
+			answered = 1;
+			break;
+		}
+	}
+	cdr = head_cdr;
 
 	for (; cdr ; cdr = cdr->next) {
+
 		/* For people, who don't want to see unanswered single-channel events */
 		if (!ast_test_flag(&mod_cfg->general->settings, CDR_UNANSWERED) &&
 				cdr->disposition < AST_CDR_ANSWERED &&
@@ -3650,6 +3661,14 @@ static void post_cdr(struct ast_cdr *cdr
 			ast_debug(1, "Skipping CDR for %s since we weren't answered\n", cdr->channel);
 			continue;
 		}
+        	/*There are multiple extensions configured in asterisk, so for one call there will be multiple cdr entries.
+	        The requirement is to log only one entry to Master.csv for one call.
+        	If the call was answered by one of the extensions that should be logged but not the entries for other extensions.
+	        */
+        	if ((answered == 1) && (cdr->disposition != AST_CDR_ANSWERED)) {
+	            ast_debug(1, "Skipping unanswered CDR for %s \n", cdr->channel);
+        	    continue;
+        	}
 
 		/*
 		 * When Asterisk receives an SIP-INVITE from an unknown source, a CDR is generated. The workaround below
@@ -3682,6 +3701,11 @@ static void post_cdr(struct ast_cdr *cdr
 			}
 		}
 		AST_RWLIST_UNLOCK(&be_list);
+		/*If the call was unanswered, only one entry should be logged into the Master.csv and rest should be ignored.*/
+		if (answered == 0) {
+			ast_debug(1, "Skip rest of the CDR entries as call was not answered\n");
+			break;
+		}
 	}
 	ao2_cleanup(mod_cfg);
 }
