diff --git a/scripts/ipping b/scripts/ipping
index f7613d1..bf933c4 100755
--- a/scripts/ipping
+++ b/scripts/ipping
@@ -122,6 +122,7 @@ ipping_launch() {
 		[ "$min_time_det" -lt $min_det ] && min_det=$min_time_det
 		[ "$max_time_det" -gt $max_det ] && max_det=$max_time_det
 		avg_time_sum_det=$((avg_time_sum_det+avg_time_det))
+		[ $idx -ne ${nbr_of_rep} ] && sleep 1 # prevent ping flood
 	done
 
 	failure_count=$((nbr_of_rep-success_count))
