diff --git a/dmentry.c b/dmentry.c
index 34023ff914d5e047b8938cfee81f3c8e0c260a1e..1da1d837d7fba045732bd1a588b2457c4156ee67 100644
--- a/dmentry.c
+++ b/dmentry.c
@@ -287,16 +287,14 @@ int dm_entry_param_method(struct dmctx *ctx, int cmd, char *inparam, char *arg1,
 		case CMD_ADD_OBJECT:
 			fault = dm_entry_add_object(ctx);
 			if (!fault) {
-				dmuci_set_value_varstate("cwmp", "cpe", "ParameterKey", arg1 ? arg1 : "");
-				dmuci_save_package_varstate("cwmp");
+				dmuci_set_value("cwmp", "cpe", "ParameterKey", arg1 ? arg1 : "");
 				dmuci_change_packages(&head_package_change);
 			}
 			break;
 		case CMD_DEL_OBJECT:
 			fault = dm_entry_delete_object(ctx);
 			if (!fault) {
-				dmuci_set_value_varstate("cwmp", "cpe", "ParameterKey", arg1 ? arg1 : "");
-				dmuci_save_package_varstate("cwmp");
+				dmuci_set_value("cwmp", "cpe", "ParameterKey", arg1 ? arg1 : "");
 				dmuci_change_packages(&head_package_change);
 			}
 			break;
@@ -347,8 +345,7 @@ int dm_entry_apply(struct dmctx *ctx, int cmd, char *arg1)
 				set_success = true;
 			}
 			if (!fault && set_success == true) {
-				dmuci_set_value_varstate("cwmp", "cpe", "ParameterKey", arg1 ? arg1 : "");
-				dmuci_save_package_varstate("cwmp");
+				dmuci_set_value("cwmp", "cpe", "ParameterKey", arg1 ? arg1 : "");
 				dmuci_change_packages(&head_package_change);
 				dmuci_save();
 			}
@@ -440,7 +437,6 @@ int dm_entry_manage_services(struct blob_buf *bb, bool restart)
 	}
 	blobmsg_close_array(bb, arr);
 
-	dmuci_commit_package_varstate("cwmp");
 	free_all_list_package_change(&head_package_change);
 	return 0;
 }
@@ -455,7 +451,6 @@ int dm_entry_restart_services(void)
 		dmubus_call_set("uci", "commit", UBUS_ARGS{{"config", pc->package, String}}, 1);
 	}
 
-	dmuci_commit_package_varstate("cwmp");
 	free_all_list_package_change(&head_package_change);
 
 	return 0;
@@ -470,7 +465,6 @@ int dm_entry_revert_changes(void)
 	list_for_each_entry(pc, &head_package_change, list) {
 		dmubus_call_set("uci", "revert", UBUS_ARGS{{"config", pc->package, String}}, 1);
 	}
-	dmuci_revert_package_varstate("cwmp");
 	free_all_list_package_change(&head_package_change);
 
 	return 0;
diff --git a/dmtree/tr181/managementserver.c b/dmtree/tr181/managementserver.c
index b24f18c622d39fb8189d580009a91380c918db9d..62525b1973dd4111a2aa22d93d247d2f6d7fd508 100644
--- a/dmtree/tr181/managementserver.c
+++ b/dmtree/tr181/managementserver.c
@@ -338,7 +338,7 @@ static int set_management_server_delay_reboot(char *refparam, struct dmctx *ctx,
 /*#Device.ManagementServer.ParameterKey!UCI:cwmp/acs,acs/ParameterKey*/
 static int get_management_server_key(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
 {
-	dmuci_get_option_value_string_varstate("cwmp", "cpe", "ParameterKey", value);
+	dmuci_get_option_value_string("cwmp", "cpe", "ParameterKey", value);
 	return 0;	
 }
 
