diff --git a/dmdiagnostics.c b/dmdiagnostics.c
index 6c93565..5059af0 100644
--- a/dmdiagnostics.c
+++ b/dmdiagnostics.c
@@ -10,6 +10,7 @@
  */
 
 #include <stdlib.h>
+#include <fcntl.h>
 #include <curl/curl.h>
 
 #ifdef LOPENSSL
@@ -608,7 +662,7 @@ int bbf_fw_image_download(const char *url, const char *auto_activate, const char
 		const char *file_size, const char *checksum_algorithm, const char *checksum,
 		const char *bank_id, const char *command, const char *obj_path, const char *commandKey)
 {
-	char fw_image_path[256] = "/tmp/firmware-XXXXXX";
+	char fw_image_path[256] = "/tmp/firmware.bin";
 	json_object *json_obj = NULL;
 	bool activate = false;
 	int res = 0;
@@ -619,7 +673,7 @@ int bbf_fw_image_download(const char *url, const char *auto_activate, const char
 		goto end;
 	}
 
-	res = mkstemp(fw_image_path);
+	res = open(fw_image_path, O_CREAT | O_EXCL, 0600); // open with read plus write permissions for owner only
 	if (res == -1) {
 		goto end;
 	} else {
