--- a/libbbf_api/dmjson.c
+++ b/libbbf_api/dmjson.c
@@ -128,12 +128,16 @@ char *____dmjson_get_value_in_array_idx(
 {
 	json_object *jobj = NULL;
 	char *value = NULL;
+	int json_arr_len;
 
 	if (arrobj == NULL || *arrobj == NULL) {
 		jobj = dmjson_select_obj(mainjobj, argv);
 		if (arrobj)
 			*arrobj = jobj;
 		if (jobj && json_object_get_type(jobj) == json_type_array) {
+       		json_arr_len = json_object_array_length(jobj);
+       		if (json_arr_len == 0)
+               	    return NULL;
 			jobj = json_object_array_get_idx(jobj, index);
 			if (jobj == NULL)
 				return NULL;
