--- a/dmtree/tr104/servicesvoiceservicedect.c
+++ b/dmtree/tr104/servicesvoiceservicedect.c
@@ -254,6 +254,249 @@
 	return 0;
 }
 
+static int del_ServicesVoiceServiceDECTBase_Contact(char *refparam, struct dmctx *ctx, void *data, char *instance, unsigned char del_action)
+{
+	json_object *res = NULL;
+
+	switch (del_action) {
+	case DEL_INST:
+		dmubus_call("dect", "contact", UBUS_ARGS{{"operation", "delete", String}, {"entry_id", instance, Integer}}, 2, &res);
+		break;
+	case DEL_ALL:
+		dmubus_call("dect", "contact", UBUS_ARGS{{"operation", "delete_all", String}}, 1, &res);
+		break;
+	}
+	return 0;
+}
+
+static int add_ServicesVoiceServiceDECTBase_Contact(char *refparam, struct dmctx *ctx, void *data, char **instance)
+{
+	json_object *res = NULL;
+
+	dmubus_call("dect", "contact", UBUS_ARGS{{"operation", "add", String}}, 1, &res);
+
+	if (res == NULL)
+		return FAULT_9002;
+
+	dmasprintf(instance, "%s", dmjson_get_value(res, 1, "entry_id"));
+	return 0;
+}
+
+/*#Device.Services.VoiceService.{i}.DECT.Base.{i}.BBF_VENDOR_PREFIX_ContactNumberOfEntries!UBUS:dect/contact/None,None/contact*/
+static int get_ServicesVoiceServiceDECTBase_ContactNumOfEntries(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
+{
+	json_object *res = NULL, *contact;
+
+	dmubus_call("dect", "contact", UBUS_ARGS{{"operation", "list", String}}, 1, &res);
+	DM_ASSERT(res, *value = "0");
+
+	json_object_object_get_ex(res, "contact", &contact);
+	dmasprintf(value, "%d", (contact) ? json_object_array_length(contact) : 0);
+
+	return 0;
+}
+
+/*#Device.Services.VoiceService.{i}.DECT.Base.{i}.BBF_VENDOR_PREFIX_Contact.{i}.!UBUS:dect/contact/None,None/contact*/
+static int browseServicesVoiceServiceDECTBaseContactInst(struct dmctx *dmctx, DMNODE *parent_node, void *prev_data, char *prev_instance)
+{
+	int i = 0;
+	char *inst = NULL;
+	json_object *res = NULL, *obj = NULL, *arrobj = NULL;
+
+	dmubus_call("dect", "contact", UBUS_ARGS{{"operation", "list", String}}, 1, &res);
+	if (res == NULL) {
+		return 0;
+	}
+
+	dmjson_foreach_obj_in_array(res, arrobj, obj, i, 1, "contact") {
+		char *entry_id = dmjson_get_value(obj, 1, "entry_id");
+
+		inst = handle_instance_without_section(dmctx, parent_node, atoi(entry_id));
+
+		if (DM_LINK_INST_OBJ(dmctx, parent_node, (void *)obj, inst) == DM_STOP)
+			break;
+	}
+
+	return 0;
+}
+
+/*#Device.Services.VoiceService.{i}.DECT.Base.{i}.BBF_VENDOR_PREFIX_Contact.{i}.FirstName!UBUS:dect/contact//contact[@i-1].first_name*/
+static int get_ServicesVoiceServiceDECTBaseContact_FirstName(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
+{
+	*value = dmjson_get_value((json_object *)data, 1, "first_name");
+
+	return 0;
+}
+
+/*#Device.Services.VoiceService.{i}.DECT.Base.{i}.BBF_VENDOR_PREFIX_Contact.{i}.LastName!UBUS:dect/contact//contact[@i-1].last_name*/
+static int get_ServicesVoiceServiceDECTBaseContact_LastName(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
+{
+	*value = dmjson_get_value((json_object *)data, 1, "last_name");
+
+	return 0;
+}
+
+/*#Device.Services.VoiceService.{i}.DECT.Base.{i}.BBF_VENDOR_PREFIX_Contact.{i}.Home!UBUS:dect/contact//contact[@i-1].home*/
+static int get_ServicesVoiceServiceDECTBaseContact_Home(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
+{
+	*value = dmjson_get_value((json_object *)data, 1, "home");
+
+	return 0;
+}
+
+/*#Device.Services.VoiceService.{i}.DECT.Base.{i}.BBF_VENDOR_PREFIX_Contact.{i}.Work!UBUS:dect/contact//contact[@i-1].work*/
+static int get_ServicesVoiceServiceDECTBaseContact_Work(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
+{
+	*value = dmjson_get_value((json_object *)data, 1, "work");
+
+	return 0;
+}
+
+/*#Device.Services.VoiceService.{i}.DECT.Base.{i}.BBF_VENDOR_PREFIX_Contact.{i}.Mobile!UBUS:dect/contact//contact[@i-1].mobile*/
+static int get_ServicesVoiceServiceDECTBaseContact_Mobile(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
+{
+	*value = dmjson_get_value((json_object *)data, 1, "mobile");
+
+	return 0;
+}
+
+/*#Device.Services.VoiceService.{i}.DECT.Base.{i}.BBF_VENDOR_PREFIX_Contact.{i}.Melody!UBUS:dect/contact//contact[@i-1].melody*/
+static int get_ServicesVoiceServiceDECTBaseContact_Melody(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
+{
+	*value = dmjson_get_value((json_object *)data, 1, "ring_melody");
+
+	return 0;
+}
+
+/*#Device.Services.VoiceService.{i}.DECT.Base.{i}.BBF_VENDOR_PREFIX_Contact.{i}.LineId!UBUS:dect/contact//contact[@i-1].line_id*/
+static int get_ServicesVoiceServiceDECTBaseContact_LineId(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
+{
+	*value = dmjson_get_value((json_object *)data, 1, "line_id");
+
+	return 0;
+}
+
+/*#Device.Services.VoiceService.{i}.DECT.Base.{i}.BBF_VENDOR_PREFIX_Contact.{i}.Alias!UBUS:dect/contact//contact[@i-1].entry_id*/
+static int get_ServicesVoiceServiceDECTBaseContact_Alias(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
+{
+	struct uci_section *s = NULL;
+
+	uci_path_foreach_option_eq(bbfdm, "dmmap", "dect", "contact_inst", instance, s) {
+		dmuci_get_value_by_section_string(s, "alias", value);
+		break;
+	}
+
+	if ((*value)[0] == '\0')
+		dmasprintf(value, "cpe-%s", dmjson_get_value((json_object *)data, 1, "entry_id"));
+
+	return 0;
+}
+
+static int set_ServicesVoiceServiceDECTBaseContact_Alias(char *refparam, struct dmctx *ctx, void *data, char *instance, char *value, int action)
+{
+	struct uci_section *s = NULL, *dmmap = NULL;
+
+	switch (action)	{
+		case VALUECHECK:
+			if (dm_validate_string(value, -1, 64, NULL, NULL))
+				return FAULT_9007;
+			break;
+		case VALUESET:
+			uci_path_foreach_option_eq(bbfdm, "dmmap", "dect", "contact_inst", instance, s) {
+				dmuci_set_value_by_section_bbfdm(s, "alias", value);
+				return 0;
+			}
+			dmuci_add_section_bbfdm("dmmap", "dect", &dmmap);
+			dmuci_set_value_by_section(dmmap, "contact_inst", instance);
+			dmuci_set_value_by_section(dmmap, "alias", value);
+			break;
+	}
+	return 0;
+}
+
+static int set_dect_contact_name(char *refparam, struct dmctx *ctx, void *data, char *instance, char *value, int action, char *field)
+{
+	switch (action)	{
+		case VALUECHECK:
+			if (dm_validate_string(value, 1, 16, NULL, NULL))
+				return FAULT_9007;
+			break;
+		case VALUESET:
+			dmubus_call_set("dect", "contact", UBUS_ARGS{{"operation", "modify", String}, {"entry_id", instance, Integer}, {field, value, String}}, 3);
+			break;
+	}
+
+	return 0;
+}
+
+static int set_dect_contact_number(char *refparam, struct dmctx *ctx, void *data, char *instance, char *value, int action, char *field)
+{
+	char *DECTNumber[] = {"^[0-9P#\\*]+$", NULL};
+
+	switch (action)	{
+		case VALUECHECK:
+			if (dm_validate_string(value, 1, 24, NULL, DECTNumber))
+				return FAULT_9007;
+			break;
+		case VALUESET:
+			dmubus_call_set("dect", "contact", UBUS_ARGS{{"operation", "modify", String}, {"entry_id", instance, Integer}, {field, value, String}}, 3);
+			break;
+	}
+
+	return 0;
+}
+
+static int set_dect_contact_uint(char *refparam, struct dmctx *ctx, void *data, char *instance, char *value, int action,
+	       			 char *field, struct range_args r_args[], int r_args_size)
+{
+	switch (action)	{
+		case VALUECHECK:
+			if (dm_validate_unsignedInt(value, r_args, r_args_size))
+				return FAULT_9007;
+			break;
+		case VALUESET:
+			dmubus_call_set("dect", "contact", UBUS_ARGS{{"operation", "modify", String}, {"entry_id", instance, Integer}, {field, value, Integer}}, 3);
+			break;
+	}
+
+	return 0;
+}
+
+static int set_ServicesVoiceServiceDECTBaseContact_FirstName(char *refparam, struct dmctx *ctx, void *data, char *instance, char *value, int action)
+{
+	return set_dect_contact_name(refparam, ctx, data, instance, value, action, "first_name");
+}
+
+static int set_ServicesVoiceServiceDECTBaseContact_LastName(char *refparam, struct dmctx *ctx, void *data, char *instance, char *value, int action)
+{
+	return set_dect_contact_name(refparam, ctx, data, instance, value, action, "last_name");
+}
+
+static int set_ServicesVoiceServiceDECTBaseContact_Home(char *refparam, struct dmctx *ctx, void *data, char *instance, char *value, int action)
+{
+	return set_dect_contact_number(refparam, ctx, data, instance, value, action, "home");
+}
+
+static int set_ServicesVoiceServiceDECTBaseContact_Work(char *refparam, struct dmctx *ctx, void *data, char *instance, char *value, int action)
+{
+	return set_dect_contact_number(refparam, ctx, data, instance, value, action, "work");
+}
+
+static int set_ServicesVoiceServiceDECTBaseContact_Mobile(char *refparam, struct dmctx *ctx, void *data, char *instance, char *value, int action)
+{
+	return set_dect_contact_number(refparam, ctx, data, instance, value, action, "mobile");
+}
+
+static int set_ServicesVoiceServiceDECTBaseContact_Melody(char *refparam, struct dmctx *ctx, void *data, char *instance, char *value, int action)
+{
+	return set_dect_contact_uint(refparam, ctx, data, instance, value, action, "ring_melody", RANGE_ARGS{{"0","15"}}, 1);
+}
+
+static int set_ServicesVoiceServiceDECTBaseContact_LineId(char *refparam, struct dmctx *ctx, void *data, char *instance, char *value, int action)
+{
+	return set_dect_contact_uint(refparam, ctx, data, instance, value, action, "line_id", RANGE_ARGS{{"0","100"}}, 1);
+}
+
 /*#Device.Services.VoiceService.{i}.DECT.Portable.{i}.Status!UBUS:dect/status//handsets[@i-1].status*/
 static int get_ServicesVoiceServiceDECTPortable_Status(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
 {
@@ -415,7 +658,7 @@
 /* *** Device.Services.VoiceService.{i}.DECT. *** */
 DMOBJ tServicesVoiceServiceDECTObj[] = {
 /* OBJ, permission, addobj, delobj, checkdep, browseinstobj, nextdynamicobj, dynamicleaf, nextobj, leaf, linker, bbfdm_type, uniqueKeys*/
-{"Base", &DMREAD, NULL, NULL, NULL, browseServicesVoiceServiceDECTBaseInst, NULL, NULL, NULL, tServicesVoiceServiceDECTBaseParams, NULL, BBFDM_BOTH, LIST_KEY{"RFPI", "Name", "Alias", NULL}},
+{"Base", &DMREAD, NULL, NULL, NULL, browseServicesVoiceServiceDECTBaseInst, NULL, NULL, tServicesVoiceServiceDECTBaseObj, tServicesVoiceServiceDECTBaseParams, NULL, BBFDM_BOTH, LIST_KEY{"RFPI", "Name", "Alias", NULL}},
 {"Portable", &DMREAD, NULL, NULL, NULL, browseServicesVoiceServiceDECTPortableInst, NULL, NULL, NULL, tServicesVoiceServiceDECTPortableParams, get_voice_service_dect_linker, BBFDM_BOTH, LIST_KEY{"IPEI", "Alias", NULL}},
 {0}
 };
@@ -439,6 +682,27 @@
 {"FirmwareVersion", &DMREAD, DMT_STRING, get_ServicesVoiceServiceDECTBase_FirmwareVersion, NULL, BBFDM_BOTH},
 {"EepromVersion", &DMREAD, DMT_STRING, get_ServicesVoiceServiceDECTBase_EepromVersion, NULL, BBFDM_BOTH},
 {"HardwareVersion", &DMREAD, DMT_STRING, get_ServicesVoiceServiceDECTBase_HardwareVersion, NULL, BBFDM_BOTH},
+{BBF_VENDOR_PREFIX"ContactNumberOfEntries", &DMREAD, DMT_UNINT, get_ServicesVoiceServiceDECTBase_ContactNumOfEntries, NULL, BBFDM_USP},
+{0}
+};
+
+/* *** Device.Services.VoiceService.{i}.DECT.Base.{i} *** */
+DMOBJ tServicesVoiceServiceDECTBaseObj[] = {
+/* OBJ, permission, addobj, delobj, checkdep, browseinstobj, nextdynamicobj, dynamicleaf, nextobj, leaf, linker, bbfdm_type, uniqueKeys*/
+{BBF_VENDOR_PREFIX"Contact", &DMWRITE, add_ServicesVoiceServiceDECTBase_Contact, del_ServicesVoiceServiceDECTBase_Contact, NULL, browseServicesVoiceServiceDECTBaseContactInst, NULL, NULL, NULL, tServicesVoiceServiceDECTBaseContactParams, NULL, BBFDM_USP},
+{0}
+};
+
+DMLEAF tServicesVoiceServiceDECTBaseContactParams[] = {
+/* PARAM, permission, type, getvalue, setvalue, bbfdm_type*/
+{"FirstName", &DMWRITE, DMT_STRING, get_ServicesVoiceServiceDECTBaseContact_FirstName, set_ServicesVoiceServiceDECTBaseContact_FirstName, BBFDM_USP},
+{"LastName", &DMWRITE, DMT_STRING, get_ServicesVoiceServiceDECTBaseContact_LastName, set_ServicesVoiceServiceDECTBaseContact_LastName, BBFDM_USP},
+{"Home", &DMWRITE, DMT_STRING, get_ServicesVoiceServiceDECTBaseContact_Home, set_ServicesVoiceServiceDECTBaseContact_Home, BBFDM_USP},
+{"Work", &DMWRITE, DMT_STRING, get_ServicesVoiceServiceDECTBaseContact_Work, set_ServicesVoiceServiceDECTBaseContact_Work, BBFDM_USP},
+{"Mobile", &DMWRITE, DMT_STRING, get_ServicesVoiceServiceDECTBaseContact_Mobile, set_ServicesVoiceServiceDECTBaseContact_Mobile, BBFDM_USP},
+{"Melody", &DMWRITE, DMT_UNINT, get_ServicesVoiceServiceDECTBaseContact_Melody, set_ServicesVoiceServiceDECTBaseContact_Melody, BBFDM_USP},
+{"LineId", &DMWRITE, DMT_UNINT, get_ServicesVoiceServiceDECTBaseContact_LineId, set_ServicesVoiceServiceDECTBaseContact_LineId, BBFDM_USP},
+{"Alias", &DMWRITE, DMT_STRING, get_ServicesVoiceServiceDECTBaseContact_Alias, set_ServicesVoiceServiceDECTBaseContact_Alias, BBFDM_USP},
 {0}
 };
 
--- a/dmtree/tr104/servicesvoiceservicedect.h
+++ b/dmtree/tr104/servicesvoiceservicedect.h
@@ -17,6 +17,8 @@
 extern DMLEAF tServicesVoiceServiceDECTParams[];
 extern DMLEAF tServicesVoiceServiceDECTBaseParams[];
 extern DMLEAF tServicesVoiceServiceDECTPortableParams[];
+extern DMOBJ tServicesVoiceServiceDECTBaseObj[];
+extern DMLEAF tServicesVoiceServiceDECTBaseContactParams[];
 
 
 #endif //__SERVICESVOICESERVICEDECT_H
