diff --git a/libwifi/modules/broadcom/wlctrl.c b/libwifi/modules/broadcom/wlctrl.c
index db1bd3e00d269a25a55f6d8c996861481685c910..9e1d5d8c4fa21065e7ce08b230ffcd3bc8c4a58b 100644
--- a/libwifi/modules/broadcom/wlctrl.c
+++ b/libwifi/modules/broadcom/wlctrl.c
@@ -3296,6 +3296,10 @@ int bcmwl_register_event(const char *ifname, struct event_struct *ev,
 	Cmd(buf, sizeof(buf), "ebtables -t broute -D BROUTING -p 0x886c -j DROP");
 	Cmd(buf, sizeof(buf), "ebtables -t broute -I BROUTING 1 -p 0x886c -j DROP");
 
+	/* Don't pass EAPOLs to bridge */
+	Cmd(buf, sizeof(buf), "ebtables -t broute -D BROUTING -p 0x888e -j DROP");
+	Cmd(buf, sizeof(buf), "ebtables -t broute -I BROUTING 1 -p 0x888e -j DROP");
+
 	return 0;
 }
 
