Index: netifd-2023-09-19-7a58b995/bridge.c
===================================================================
--- netifd-2023-09-19-7a58b995.orig/bridge.c
+++ netifd-2023-09-19-7a58b995/bridge.c
@@ -235,7 +235,7 @@ bridge_set_local_vlan(struct bridge_stat
 	if (!vlan->local && add)
 		return;
 
-	system_bridge_vlan(bst->dev.ifname, vlan->vid, -1, add, BRVLAN_F_SELF);
+	system_bridge_vlan(bst->dev.ifname, vlan->vid, -1, add, vlan->local_flags | BRVLAN_F_SELF);
 }
 
 static void
@@ -1024,6 +1024,7 @@ bridge_dump_vlan(struct blob_buf *b, str
 
 	blobmsg_add_u32(b, "id", vlan->vid);
 	blobmsg_add_u8(b, "local", vlan->local);
+	blobmsg_add_u32(b, "local_flags", vlan->local_flags);
 
 	p = blobmsg_open_array(b, "ports");
 
Index: netifd-2023-09-19-7a58b995/device.h
===================================================================
--- netifd-2023-09-19-7a58b995.orig/device.h
+++ netifd-2023-09-19-7a58b995/device.h
@@ -315,6 +315,7 @@ struct bridge_vlan {
 
 	uint16_t vid;
 	bool local;
+	enum bridge_vlan_flags local_flags;
 	bool pending;
 };
 
Index: netifd-2023-09-19-7a58b995/config.c
===================================================================
--- netifd-2023-09-19-7a58b995.orig/config.c
+++ netifd-2023-09-19-7a58b995/config.c
@@ -312,6 +312,7 @@ config_parse_vlan(struct device *dev, st
 	enum {
 		BRVLAN_ATTR_VID,
 		BRVLAN_ATTR_LOCAL,
+		BRVLAN_ATTR_FLAGS,
 		BRVLAN_ATTR_PORTS,
 		BRVLAN_ATTR_ALIAS,
 		__BRVLAN_ATTR_MAX,
@@ -319,6 +320,7 @@ config_parse_vlan(struct device *dev, st
 	static const struct blobmsg_policy vlan_attrs[__BRVLAN_ATTR_MAX] = {
 		[BRVLAN_ATTR_VID] = { "vlan", BLOBMSG_TYPE_INT32 },
 		[BRVLAN_ATTR_LOCAL] = { "local", BLOBMSG_TYPE_BOOL },
+		[BRVLAN_ATTR_FLAGS] = { "flags", BLOBMSG_TYPE_STRING },
 		[BRVLAN_ATTR_PORTS] = { "ports", BLOBMSG_TYPE_ARRAY },
 		[BRVLAN_ATTR_ALIAS] = { "alias", BLOBMSG_TYPE_ARRAY },
 	};
@@ -338,6 +340,7 @@ config_parse_vlan(struct device *dev, st
 	unsigned int vid;
 	const char *val;
 	char *name_buf;
+	char *local_flags;
 	int name_len = 0;
 	int n_ports = 0;
 	size_t rem;
@@ -371,6 +374,15 @@ config_parse_vlan(struct device *dev, st
 	if (tb[BRVLAN_ATTR_LOCAL])
 		vlan->local = blobmsg_get_bool(tb[BRVLAN_ATTR_LOCAL]);
 
+	if (tb[BRVLAN_ATTR_FLAGS]) {
+		local_flags = blobmsg_get_string(tb[BRVLAN_ATTR_FLAGS]);
+
+		if (strstr(local_flags, "pvid"))
+			vlan->local_flags |= BRVLAN_F_PVID;
+		if (strstr(local_flags, "untagged"))
+			vlan->local_flags |= BRVLAN_F_UNTAGGED;
+	}
+
 	vlan->n_ports = n_ports;
 	vlan->ports = port = (struct bridge_vlan_port *)&vlan[1];
 	INIT_LIST_HEAD(&vlan->hotplug_ports);
