diff -Nur libbbfdm-6.4.27.orig/dmtree/tr104/servicesvoiceservicecalllog.c libbbfdm-6.4.27/dmtree/tr104/servicesvoiceservicecalllog.c
--- a/dmtree/tr104/servicesvoiceservicecalllog.c	2022-05-09 21:15:05.836477143 +0100
+++ b/dmtree/tr104/servicesvoiceservicecalllog.c	2022-05-09 21:24:50.171981035 +0100
@@ -278,28 +278,42 @@
 /* *** Device.Services.VoiceService.{i}.CallLog.{i}.Session.{i}.Destination.DSP. *** */
 DMOBJ tServicesVoiceServiceCallLogSessionDestinationDSPObj[] = {
 /* OBJ, permission, addobj, delobj, checkdep, browseinstobj, nextdynamicobj, dynamicleaf, nextobj, leaf, linker, bbfdm_type, uniqueKeys*/
-{"ReceiveCodec", &DMREAD, NULL, NULL, NULL, NULL, NULL, NULL, NULL, tServicesVoiceServiceCallLogSessionDestinationDSPCodecParams, NULL, BBFDM_BOTH},
-{"TransmitCodec", &DMREAD, NULL, NULL, NULL, NULL, NULL, NULL, NULL, tServicesVoiceServiceCallLogSessionDestinationDSPCodecParams, NULL, BBFDM_BOTH},
+{"ReceiveCodec", &DMREAD, NULL, NULL, NULL, NULL, NULL, NULL, NULL, tServicesVoiceServiceCallLogSessionDestinationDSPReceiveCodecParams, NULL, BBFDM_BOTH},
+{"TransmitCodec", &DMREAD, NULL, NULL, NULL, NULL, NULL, NULL, NULL, tServicesVoiceServiceCallLogSessionDestinationDSPTransmitCodecParams, NULL, BBFDM_BOTH},
 {0}
 };
 
 /* *** Device.Services.VoiceService.{i}.CallLog.{i}.Session.{i}.Source.DSP. *** */
 DMOBJ tServicesVoiceServiceCallLogSessionSourceDSPObj[] = {
 /* OBJ, permission, addobj, delobj, checkdep, browseinstobj, nextdynamicobj, dynamicleaf, nextobj, leaf, linker, bbfdm_type, uniqueKeys*/
-{"ReceiveCodec", &DMREAD, NULL, NULL, NULL, NULL, NULL, NULL, NULL, tServicesVoiceServiceCallLogSessionSourceDSPCodecParams, NULL, BBFDM_BOTH},
-{"TransmitCodec", &DMREAD, NULL, NULL, NULL, NULL, NULL, NULL, NULL, tServicesVoiceServiceCallLogSessionSourceDSPCodecParams, NULL, BBFDM_BOTH},
+{"ReceiveCodec", &DMREAD, NULL, NULL, NULL, NULL, NULL, NULL, NULL, tServicesVoiceServiceCallLogSessionSourceDSPReceiveCodecParams, NULL, BBFDM_BOTH},
+{"TransmitCodec", &DMREAD, NULL, NULL, NULL, NULL, NULL, NULL, NULL, tServicesVoiceServiceCallLogSessionSourceDSPTransmitCodecParams, NULL, BBFDM_BOTH},
 {0}
 };
 
 /* *** Device.Services.VoiceService.{i}.CallLog.{i}.Session.{i}.Destination.DSP.ReceiveCodec. *** */
-DMLEAF tServicesVoiceServiceCallLogSessionDestinationDSPCodecParams[] = {
+DMLEAF tServicesVoiceServiceCallLogSessionDestinationDSPReceiveCodecParams[] = {
+/* PARAM, permission, type, getvalue, setvalue, bbfdm_type*/
+{"Codec", &DMREAD, DMT_STRING, get_ServicesVoiceServiceCallLog_SessionDSPCodec, NULL, BBFDM_BOTH},
+{0}
+};
+
+/* *** Device.Services.VoiceService.{i}.CallLog.{i}.Session.{i}.Destination.DSP.TransmitCodec. *** */
+DMLEAF tServicesVoiceServiceCallLogSessionDestinationDSPTransmitCodecParams[] = {
 /* PARAM, permission, type, getvalue, setvalue, bbfdm_type*/
 {"Codec", &DMREAD, DMT_STRING, get_ServicesVoiceServiceCallLog_SessionDSPCodec, NULL, BBFDM_BOTH},
 {0}
 };
 
 /* *** Device.Services.VoiceService.{i}.CallLog.{i}.Session.{i}.Source.DSP.ReceiveCodec. *** */
-DMLEAF tServicesVoiceServiceCallLogSessionSourceDSPCodecParams[] = {
+DMLEAF tServicesVoiceServiceCallLogSessionSourceDSPReceiveCodecParams[] = {
+/* PARAM, permission, type, getvalue, setvalue, bbfdm_type*/
+{"Codec", &DMREAD, DMT_STRING, get_ServicesVoiceServiceCallLog_SessionDSPCodec, NULL, BBFDM_BOTH},
+{0}
+};
+
+/* *** Device.Services.VoiceService.{i}.CallLog.{i}.Session.{i}.Source.DSP.TransmitCodec. *** */
+DMLEAF tServicesVoiceServiceCallLogSessionSourceDSPTransmitCodecParams[] = {
 /* PARAM, permission, type, getvalue, setvalue, bbfdm_type*/
 {"Codec", &DMREAD, DMT_STRING, get_ServicesVoiceServiceCallLog_SessionDSPCodec, NULL, BBFDM_BOTH},
 {0}
diff -Nur libbbfdm-6.4.27.orig/dmtree/tr104/servicesvoiceservicecalllog.h libbbfdm-6.4.27/dmtree/tr104/servicesvoiceservicecalllog.h
--- a/dmtree/tr104/servicesvoiceservicecalllog.h	2022-05-09 21:15:05.836477143 +0100
+++ b/dmtree/tr104/servicesvoiceservicecalllog.h	2022-05-09 21:18:01.769150088 +0100
@@ -21,8 +21,10 @@
 extern DMOBJ tServicesVoiceServiceCallLogSessionSourceObj[];
 extern DMOBJ tServicesVoiceServiceCallLogSessionDestinationDSPObj[];
 extern DMOBJ tServicesVoiceServiceCallLogSessionSourceDSPObj[];
-extern DMLEAF tServicesVoiceServiceCallLogSessionDestinationDSPCodecParams[];
-extern DMLEAF tServicesVoiceServiceCallLogSessionSourceDSPCodecParams[];
+extern DMLEAF tServicesVoiceServiceCallLogSessionDestinationDSPReceiveCodecParams[];
+extern DMLEAF tServicesVoiceServiceCallLogSessionDestinationDSPTransmitCodecParams[];
+extern DMLEAF tServicesVoiceServiceCallLogSessionSourceDSPReceiveCodecParams[];
+extern DMLEAF tServicesVoiceServiceCallLogSessionSourceDSPTransmitCodecParams[];
 extern DMLEAF tServicesVoiceServiceCallLogSessionSourceRTPParams[];
 
 #endif //__SERVICESVOICESERVICECALLLOG_H
