Index: libbbfdm-6.8.9.20/dmtree/tr181/dhcpv4.c
===================================================================
--- libbbfdm-6.8.9.20.orig/dmtree/tr181/dhcpv4.c
+++ libbbfdm-6.8.9.20/dmtree/tr181/dhcpv4.c
@@ -346,22 +346,6 @@ static void dhcp_leases_assign_to_interf
 	}
 }
 
-static bool check_dhcp_host_alias_exists(char *dhcp_interface, char *option, char *value)
-{
-	struct uci_section *s = NULL;
-	char *opt_value;
-
-	uci_path_foreach_option_eq(bbfdm, "dmmap_dhcp", "host", "dhcp", dhcp_interface, s) {
-
-		dmuci_get_value_by_section_string(s, option, &opt_value);
-
-		if (DM_STRCMP(opt_value, value) == 0)
-			return true;
-	}
-
-	return false;
-}
-
 static bool check_dhcp_host_option_exists(char *dhcp_interface, char *option, char *value)
 {
 	struct uci_section *s = NULL;
@@ -1914,24 +1898,12 @@ static int get_DHCPv4ServerPoolStaticAdd
 
 static int set_DHCPv4ServerPoolStaticAddress_Alias(char *refparam, struct dmctx *ctx, void *data, char *instance, char *value, int action)
 {
-	char *curr_alias = NULL, *alias_assigned = NULL;
-
 	switch (action) {
 		case VALUECHECK:
 			// Validate value string -> length
 			if (dm_validate_string(value, -1, 64, NULL, NULL))
 				return FAULT_9007;
 
-			// Check if alias is assigned by user
-			dmuci_get_value_by_section_string((((struct dhcp_host_args *)data)->host_sections)->dmmap_section, "dhcp_host_alias_assigned", &alias_assigned);
-			if (alias_assigned && DM_LSTRCMP(alias_assigned, "1") == 0)
-				return FAULT_9007;
-
-			// Check if alias exists
-			dmuci_get_value_by_section_string((((struct dhcp_host_args *)data)->host_sections)->dmmap_section, "dhcp_host_alias", &curr_alias);
-			if (DM_STRCMP(curr_alias, value) != 0 && check_dhcp_host_alias_exists(((struct dhcp_host_args *)data)->dhcp_interface, "dhcp_host_alias", value))
-				return FAULT_9007;
-
 			return 0;
 		case VALUESET:
 			dmuci_set_value_by_section((((struct dhcp_host_args *)data)->host_sections)->dmmap_section, "dhcp_host_alias", value);
