--- a/dmtree/tr181/wifi.c
+++ b/dmtree/tr181/wifi.c
@@ -3371,9 +3371,15 @@ static int get_access_point_associative_device_statistics_retrans_count(char *re
 static int get_wifi_access_point_status(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
 {
 	json_object *res = NULL;
-	char object[32], *status = NULL, *iface;
+	char object[32], *status = NULL, *iface, *enabled = NULL;
 
-	dmuci_get_value_by_section_string((((struct wifi_ssid_args *)data)->sections)->config_section, "device", &iface);
+	get_wifi_enable(refparam, ctx, data, instance, &enabled);
+	if (DM_LSTRCMP(enabled, "0") == 0) {
+		*value = "Disabled";
+		return 0;
+	}
+
+	dmuci_get_value_by_section_string((((struct wifi_ssid_args *)data)->sections)->config_section, "ifname", &iface);
 	snprintf(object, sizeof(object), "wifi.ap.%s", iface);
 	dmubus_call(object, "status", UBUS_ARGS{0}, 0, &res);
 	DM_ASSERT(res, *value = "Error_Misconfigured");
