Index: libbbfdm-6.8.9.20/dmtree/tr181/ieee1905.c
===================================================================
--- libbbfdm-6.8.9.20.orig/dmtree/tr181/ieee1905.c
+++ libbbfdm-6.8.9.20/dmtree/tr181/ieee1905.c
@@ -362,40 +362,7 @@ static char *get_datamodel_media_type(co
 		return (char *)media;
 }
 
-static int ubus_ieee1905_info_interface_option(const char *option, char **value)
-{
-	json_object *res = NULL, *interface_obj = NULL, *arrobj = NULL, *link_obj = NULL;
-	char *media, *mac;
-	int i = 0, j = 0, match = 0;
-	dmubus_call("ieee1905", "info", UBUS_ARGS{0}, 0, &res);
-	if (res) {
-		dmjson_foreach_obj_in_array(res, arrobj, interface_obj, i, 1, "interface") {
-			char *links = dmjson_get_value(interface_obj, 1, "num_links");
-			if (atoi(links) > 0) {
-				j  = 0;
-				while (j < atoi(links)) {
-					link_obj = dmjson_select_obj_in_array_idx(interface_obj, j, 1, "links");
-					if (link_obj) {
-						mac = dmjson_get_value(link_obj, 1, "macaddress");
-						if (mac && !strncmp(option, mac, strlen(option))) {
-							media = dmjson_get_value(link_obj, 1, "media");
-							*value = get_datamodel_media_type(media);
-							match = 1;
-							break;
-						}
-					}
-					j++;
-				}
-				if (match) {
-					break;
-				}
-			}
-		}
-	}
-	return 0;
-}
-
-static int ubus_ieee1905_info_nonneighbor_mediatype(const char *option, char **value)
+static int ubus_ieee1905_get_neighbor_localinterfacemediatype(const char *option, char **value)
 {
 	json_object *res = NULL, *device_obj = NULL, *arrobj = NULL, *interface_obj = NULL;
 
@@ -1470,7 +1437,7 @@ static int get_IEEE1905ALNetworkTopology
 static int get_IEEE1905ALNetworkTopologyIEEE1905DeviceNonIEEE1905Neighbor_LocalInterfaceMediaType(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
 {
 	char *linker = ((struct ieee1905_device_nonieee1905neighbor_args *)data)->mac_addr;
-	ubus_ieee1905_info_nonneighbor_mediatype(linker, value);
+	ubus_ieee1905_get_neighbor_localinterfacemediatype(linker, value);
 	return 0;
 }
 /*#Device.IEEE1905.AL.NetworkTopology.IEEE1905Device.{i}.NonIEEE1905Neighbor.{i}.NeighborInterfaceId!UBUS:ieee1905/info//topology.device[@i-1].non1905_neighbors[@i-1].neighbors*/
@@ -1512,7 +1479,7 @@ static int get_IEEE1905ALNetworkTopology
 static int get_IEEE1905ALNetworkTopologyIEEE1905DeviceIEEE1905Neighbor_LocalInterfaceMediaType(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
 {
 	char *linker = dmjson_get_value((json_object *)data, 1, "macaddress");
-	ubus_ieee1905_info_interface_option(linker, value);
+	ubus_ieee1905_get_neighbor_localinterfacemediatype(linker, value);
 	return 0;
 }
 
