diff --git a/dmtree/tr181/device.c b/dmtree/tr181/device.c
index 435ff3f..41a6b77 100644
--- a/dmtree/tr181/device.c
+++ b/dmtree/tr181/device.c
@@ -57,7 +57,7 @@ static int get_Device_InterfaceStackNumberOfEntries(char *refparam, struct dmctx
 
 static int get_Device_RootDataModelVersion(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
 {
-	*value = DEFAULT_DMVERSION;
+	dmasprintf(value, "%s", DEFAULT_DMVERSION);
 	return 0;
 }
 
