--- a/dmtree/tr181/wifi.c
+++ b/dmtree/tr181/wifi.c
@@ -3324,6 +3324,14 @@ static int get_WiFiAccessPointAssociatedDevice_AssociationTime(char *refparam, s
 	return 0;
 }

+static int get_WiFiAccessPointAssociatedDevice_OperatingStandard(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
+{
+	/* TODO dummy */
+	*value = "a";
+
+	return 0;
+}
+
 /*#Device.WiFi.AccessPoint.{i}.AssociatedDevice.{i}.Stats.BytesSent!UBUS:wifi.ap.@Name/stations//stations[i-1].stats.tx_total_bytes*/
 static int get_access_point_associative_device_statistics_tx_bytes(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
 {
@@ -6513,6 +6521,7 @@ DMLEAF tWiFiAccessPointAssociatedDeviceParams[] = {
 {"SignalStrength", &DMREAD, DMT_INT, get_WiFiAccessPointAssociatedDevice_SignalStrength, NULL, BBFDM_BOTH, "2.0"},
 //{"Retransmissions", &DMREAD, DMT_UNINT, get_WiFiAccessPointAssociatedDevice_Retransmissions, NULL, BBFDM_BOTH, "2.0"},
 {"AssociationTime", &DMREAD, DMT_TIME, get_WiFiAccessPointAssociatedDevice_AssociationTime, NULL, BBFDM_BOTH, "2.12"},
+{"OperatingStandard", &DMREAD, DMT_STRING, get_WiFiAccessPointAssociatedDevice_OperatingStandard, NULL, BBFDM_CWMP, "2.15"},
 {0}
 };
