--- a/dmtree/tr104/servicesvoiceservicedect.c
+++ b/dmtree/tr104/servicesvoiceservicedect.c
@@ -497,6 +497,15 @@ static int set_ServicesVoiceServiceDECTBaseContact_LineId(char *refparam, struct
 	return set_dect_contact_uint(refparam, ctx, data, instance, value, action, "line_id", RANGE_ARGS{{"0","100"}}, 1);
 }
 
+static int operate_ServicesVoiceServiceDECTBase_ContactDeleteAll(char *refparam, struct dmctx *ctx, void *data, char *instance, char *value, int action)
+{
+	json_object *res = NULL;
+
+	dmubus_call("dect", "contact", UBUS_ARGS{{"operation", "delete_all", String}}, 1, &res);
+
+	return 0;
+}
+
 /*#Device.Services.VoiceService.{i}.DECT.Portable.{i}.Status!UBUS:dect/status//handsets[@i-1].status*/
 static int get_ServicesVoiceServiceDECTPortable_Status(char *refparam, struct dmctx *ctx, void *data, char *instance, char **value)
 {
@@ -683,6 +692,7 @@ DMLEAF tServicesVoiceServiceDECTBaseParams[] = {
 {"EepromVersion", &DMREAD, DMT_STRING, get_ServicesVoiceServiceDECTBase_EepromVersion, NULL, BBFDM_BOTH},
 {"HardwareVersion", &DMREAD, DMT_STRING, get_ServicesVoiceServiceDECTBase_HardwareVersion, NULL, BBFDM_BOTH},
 {BBF_VENDOR_PREFIX"ContactNumberOfEntries", &DMREAD, DMT_UNINT, get_ServicesVoiceServiceDECTBase_ContactNumOfEntries, NULL, BBFDM_USP},
+{BBF_VENDOR_PREFIX"ContactDeleteAll()", &DMSYNC, DMT_COMMAND, NULL, operate_ServicesVoiceServiceDECTBase_ContactDeleteAll, BBFDM_USP},
 {0}
 };
 
