--- a/res/res_pjsip_session.c
+++ b/res/res_pjsip_session.c
@@ -5048,6 +5051,15 @@ static void session_inv_on_tsx_state_changed(pjsip_inv_session *inv, pjsip_trans
 		break;
 	}
 
+	if (inv->state == PJSIP_INV_STATE_CALLING && tsx->state == PJSIP_TSX_STATE_CALLING && session->channel) {
+		/* Log the IP that the INVITE been sent as CallLog.{i}.X_Vendor_SIPIPAddress */
+		const char *addr_buf;
+		char tempaddr[AST_SOCKADDR_BUFLEN];
+		addr_buf = pj_sockaddr_print(&tsx->addr, tempaddr, sizeof(tempaddr), 0);
+		ast_debug(3, "%s: INVITE sent to IP: %s \n", ast_sip_session_get_name(session), addr_buf );
+		ast_channel_sipIpAddress_set(session->channel, addr_buf, strlen(addr_buf));
+	}
+
 	if (AST_LIST_EMPTY(&session->delayed_requests)) {
 		/* No delayed request pending, so just return */
 		SCOPE_EXIT_RTN("Nothing delayed\n");
