--- a/channels/chan_brcm.c
+++ b/channels/chan_brcm.c
@@ -2310,6 +2310,10 @@ static void handle_hookflash(struct brcm
 			p->hf_detected = 0;
 			return;
 		}
+	} else if (sub->channel_state == CALLING && sub_peer->channel_state != ONHOLD){
+		//ignore the flash hook sending from endptmngr for DECT when got 183 from IVR system and interacting through DTMF
+		p->hf_detected = 0;
+		return;
 	}
 
 	if (p->dtmf_first < 0) {
